### PR TITLE
[Feature] Modify pipeline dop by running fragments for non-query-queue

### PR DESCRIPTION
**Pull request by ZiheLiu**
### Why I'm doing
When running a large number of queries with many fragments but short execution times under high concurrency, DOP=0 will create too many pipeline drivers, resulting in performance that is 4 to 5 times worse than when DOP=1.

### What I'm doing

This PR expands the previous strategy for adjusting DOP based on the number of running fragments that was used for the query queue.
- When the query queue is enabled, the leader FE adjusts the DOP and maintains the number of running drivers. (The same as before)
- When the query queue is not enabled, each FE individually adjusts the DOP and maintains the number of running drivers locally.

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/qe/scheduler/slot/GlobalSlotProvider.java [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+import java.util.stream.Collectors;
+
+/**
+ * Manage slot globally in the whole cluster when enabling query queue.
+ * <p> It is the slot manager view in the follower FEs. It receives the slot operations from
+ * {@link com.starrocks.qe.scheduler.Coordinator} and sends it to {@link SlotManager} via RPC.
+ *
+ * @see SlotManager
+ */
+public class GlobalSlotProvider implements SlotProvider {
```
Comment (ZiheLiu) [inline fe/fe-core/src/main/java/com/starrocks/qe/scheduler/slot/GlobalSlotProvider.java:48]: This file is changed nothing but renamed from `SlotProvider` to `GlobalSlotProvider`.

File: fe/fe-core/src/main/java/com/starrocks/qe/scheduler/slot/PipelineDriverAllocator.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+            numAllocatedDrivers.getAndAdd(-slot.getNumDrivers());
+        }
+    }
+
+    @VisibleForTesting
+    int getNumAllocatedDrivers() {
+        return numAllocatedDrivers.get();
+    }
+
+    private void allocateInBatch() {
```
Comment (satanson) [inline fe/fe-core/src/main/java/com/starrocks/qe/scheduler/slot/PipelineDriverAllocator.java:113]: This implementation looks like WriteBatch merging in leveldb.

File: fe/fe-core/src/main/java/com/starrocks/qe/scheduler/slot/PipelineDriverAllocator.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        if (numFragments <= 0) {
+            return 1;
+        }
+
+        // Calculate DOP by driverHighWater.
+        final int hardLimit = GlobalVariable.getQueryQueueDriverHighWater();
+        int dop = calculateDopByLimit(curNumAllocatedDrivers, numFragments, hardLimit);
+
+        if (dop <= 1) {
+            return dop;
```
Comment (satanson) [inline fe/fe-core/src/main/java/com/starrocks/qe/scheduler/slot/PipelineDriverAllocator.java:146]: here dop is always 1?  it seems returning 1 is more clear.

File: fe/fe-core/src/main/java/com/starrocks/qe/scheduler/slot/PipelineDriverAllocator.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+        return dop;
+    }
+
+    private int calculateDopByLimit(int curNumAllocatedDrivers, int numFragments, int limit) {
+        if (curNumAllocatedDrivers + numFragments >= limit) {
+            return 1;
+        }
+
+        int dop = (limit - curNumAllocatedDrivers) / numFragments;
```
Comment (satanson) [inline fe/fe-core/src/main/java/com/starrocks/qe/scheduler/slot/PipelineDriverAllocator.java:177]: it seems that here LWM should be taken into consideration.

File: fe/fe-core/src/main/java/com/starrocks/qe/scheduler/slot/PipelineDriverAllocator.java (hunk 2 above)
Comment (ZiheLiu) [reply fe/fe-core/src/main/java/com/starrocks/qe/scheduler/slot/PipelineDriverAllocator.java:113]: Yes, inspiration from leveldb.

**Review by Seaven — APPROVED** [review]

**Review by liuyehcf — APPROVED** [review]

**Pull request merged**